File: VIF_git/GSDS-Qual/LeetCode/week1/Linked_List/Odd_Even_Linked_List/sunghyun.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def oddEvenList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """

        oddlist = ListNode(0)
        evenlist = ListNode(0)
        
        oddhead = oddlist
        evenhead = evenlist

        while(head):
            oddlist.next = head
            evenlist.next = head.next
            
            oddlist = oddlist.next
            evenlist = evenlist.next
            
            if (evenlist):
                head = head.next.next
            else:
                head = None
        
        oddlist.next = evenhead.next

        return oddhead.next

        

File: VIF_git/GSDS-Qual/LeetCode/week1/Linked_List/Odd_Even_Linked_List/sunghyun.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {
        ListNode odd = ListNode();
        ListNode even = ListNode();
        
        ListNode* oddlist = &odd;
        ListNode* evenlist = &even;

        ListNode* oddhead = oddlist;
        ListNode* evenhead = evenlist;

        while(head){
            oddlist->next = head;
            evenlist->next = head->next;

            oddlist = oddlist->next;
            evenlist = evenlist->next;

            if(evenlist){
                head = head->next->next;
            }
            else{
                head = nullptr;
            }
        }

        oddlist->next = evenhead->next;

        return oddhead->next;

        
    }
};